pkg/kubelet/cm/cgroup_manager_linux.go

```diff
@@ -53,74 +53,71 @@ const (
 var hugePageSizeList = []string{"B", "kB", "MB", "GB", "TB", "PB"}
 
-// ConvertCgroupNameToSystemd converts the internal cgroup name to a systemd name.
-// For example, the name /Burstable/pod_123-456 becomes Burstable-pod_123_456.slice
-// If outputToCgroupFs is true, it expands the systemd name into the cgroupfs form.
-// For example, it will return /Burstable.slice/Burstable-pod_123_456.slice in above scenario.
-func ConvertCgroupNameToSystemd(cgroupName CgroupName, outputToCgroupFs bool) string {
-	name := string(cgroupName)
-	result := ""
-	if name != "" && name != "/" {
-		parts := strings.Split(name, "/")
-		results := []string{}
-		for _, part := range parts {
-			// ignore leading stuff
-			if part == "" {
-				continue
-			}
-			// detect if we are given a systemd style name.
-			// if so, we do not want to do double encoding.
-			if IsSystemdStyleName(part) {
-				part = strings.TrimSuffix(part, systemdSuffix)
-				separatorIndex := strings.LastIndex(part, "-")
-				if separatorIndex >= 0 && separatorIndex < len(part) {
-					part = part[separatorIndex+1:]
-				}
-			} else {
-				// systemd treats - as a step in the hierarchy, we convert all - to _
-				part = strings.Replace(part, "-", "_", -1)
-			}
-			results = append(results, part)
+var RootCgroupName = CgroupName([]string{})
+
+// NewCgroupName composes a new cgroup name.
+// Use RootCgroupName as base to start at the root.
+// This function does some basic check for invalid characters at the name.
+func NewCgroupName(base CgroupName, components ...string) CgroupName {
+	for _, component := range components {
+		// Forbit using "_" in internal names. When remapping internal
+		// names to systemd cgroup driver, we want to remap "-" => "_",
+		// so we forbid "_" so that we can always revert the mapping.
+		if strings.Contains(component, "/") || strings.Contains(component, "_") {
+			panic(fmt.Errorf("Invalid character in component [%s] of CgroupName", component))
 		}
-		// each part is appended with systemd style -
-		result = strings.Join(results, "-")
-	} else {
-		// root converts to -
-		result = "-"
 	}
-	// always have a .slice suffix
-	if !IsSystemdStyleName(result) {
-		result = result + systemdSuffix
+	return CgroupName(append(base, components...))
+}
+
+// cgroupName.ToSystemd converts the internal cgroup name to a systemd name.
+// For example, the name {"kubepods", "burstable", "pod1234-abcd-5678-efgh"} becomes
+// "/kubepods.slice/kubepods-burstable.slice/kubepods-burstable-pod1234_abcd_5678_efgh.slice"
+// This function always expands the systemd name into the cgroupfs form. If only
+// the last part is needed, use path.Base(...) on it to discard the rest.
+func (cgroupName CgroupName) ToSystemd() string {
+	if len(cgroupName) == 0 || (len(cgroupName) == 1 && cgroupName[0] == "") {
+		return "/"
+	}
+	newparts := []string{}
+	for _, part := range cgroupName {
+		part = strings.Replace(part, "-", "_", -1)
```

Comment:
Suggestion: make `_` (and `-`) a constant since it's used multiple times in this file.

Reply:
So I started that:

```golang
        // systemdSliceSeparator is the character "-" used to separate slice
        // names in a hierarchy in systemd cgroups.
        systemdSliceSeparator = "-"
        // escapedCgroupSeparator is the character "_" which we map to when
        // we find a dash ("-") in the cgroup name.  We want to allow dashes
        // in the internal names (for pod UUIDs) but we can't allow them
        // in systemd slice names since they conflict with the separator.
        escapedCgroupSeparator = "_"
```

But in the end I thought it was much harder to figure out what was happening...

Also, I have this code `strings.Contains(component, "/")`, so now do I need a constant for "/" as well?

I think the suggestion to factor out `strings.Replace(part, "_", "-", -1)` is really good in that it's easy to name that in a way that improves readability, but constants for "_" and "-" don't help much, unless they're called "dash" and "underscore", but then why bother?

(My 2c... If you think constants really improve this, reply here and I'll go that route.)